staking multiple nominations

4meta5 opened the pull request:
### What does it do?

- [x] The previous design only allowed one validator per nominator. This PR replaces that constraint with a new one that allows `T::MaxValidatorsPerNominator` validators per nominator, such that `each nomination > T::MinNomination` and `total nominated > T::MinNominatorStk`.
- [x] update chainspecs with separate input param  `stakers: Vec<(AccountId, Option<AccountId>, Balance)>`. For validators, it is `(AccountId, None, Balance)` and for nominators it is `(AccountId, Some(ValidatorAccountId), Balance)`. This logic can be found in the `stake` pallet genesis `build()` impl.
- [x] validator candidates may bond more/less
- [x] nominators may bond more/less
- [x] `switch_nomination` runtime method

### What important points reviewers should know?

Here are some high-level [docs](https://meta5.world/stake-docs), which should make the code easier to read/review.

### Is there something left for follow-up PRs?

enforcing inflation schedule, which was an original goal of this PR

### What alternative implementations were considered?

### Are there relevant PRs or issues in other repositories (Substrate, Polkadot, Frontier, Cumulus)?

 #179 

### What value does it bring to the blockchain users?

## Checklist

- Does it require a purge of the network? No
- [x] You bumped the runtime version if there are breaking changes in the **runtime** ?
- Does it require changes in documentation/tutorials ? No

4meta5 commented on pallets/stake/src/lib.rs:
```diff
@@ -111,17 +111,17 @@ impl<A, B: HasCompact> Into<IndividualExposure<A, B>> for Bond<A, B> {
 #[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
 /// The activity status of the validator
-pub enum ValidatorStatus<BlockNumber> {
+pub enum ValidatorStatus {
```
`ValidatorStatus` used to be generic over the inner variant of `Leaving`, but it was only ever instantiated with the `RoundIndex` type. So, the generic was removed and the inner variant of `ValidatorStatus::Leaving` was defined to always be `RoundIndex`. Earlier in this PR, something similar was done to remove RoundIndex as a generic parameter  from `CandidateState`.

4meta5 commented on pallets/stake/src/lib.rs:
```diff
@@ -308,15 +548,14 @@ decl_storage! {
 					)
 				} else {
-					<Module<T>>::join_candidates(
+					<Module<T>>::join_candidates_no_fee(
```
This changes the default fee for validators registered at genesis to 0%. It was previously 2%.

We could also add a `DefaultValidatorFee` config constant and use that here with the old `join_candidates` method. Then, there's no need to create the `join_candidates_no_fee` runtime method, but it does add a config constant and there are a lot already.

4meta5 replied:
Never mind, changed this so didn't add a new runtime method. The default fee was still changed to 0% for validators registered at genesis. is this a suitable default?

4meta5 commented on runtime/src/lib.rs:
```diff
@@ -345,9 +347,11 @@ impl stake::Config for Runtime {
 	type IssuancePerRound = IssuancePerRound;
 	type MaxFee = MaxFee;
 	type MinValidatorStk = MinValidatorStk;
+	type MinNomination = MinNominatorStk;
```
This sets the minimum nomination as equal to `MinNominatorStk = 5 * GLMR`, but it could be set at some value less than `MinNominatorStk`. These limits exist to prevent trivial nominations from abusing runtime storage.

4meta5 commented:
TODO: update polkadot-js types based on changes to Rust types

JoshOrndorff commented on pallets/stake/src/mock.rs:
```diff
@@ -193,7 +197,73 @@ pub fn genesis2() -> sp_io::TestExternalities {
+	ext
+}
+
+pub(crate) fn genesis4() -> sp_io::TestExternalities {
```
I think four is the point at which they start needing better names.

JoshOrndorff reviewed (commented):
The notes are awesome. Can we get those as doc comments or something?

What does it mean to go offline?

Why do we need seperate bond_more (same question about bond_less) methods for validators and nominators? Maybe you could have a single `fn bond_more(origin, candidate, more: BalanceOf<T>)` and if the candidate is the same as the caller then treat it as a candidate bonding more. Or use an option like you did for the genesis config.

Design question: Can one account be a validator and also a nominator? If so can the account self nominate?

Concern: This pallet just ticked past 1k lines code, and some parts are deeply nested. I'm reassured by the thorough test coverage. But I wonder how this footprint compares to eg pallet staking.

4meta5 commented:
> The notes are awesome. Can we get those as doc comments or something?

Thanks, the reason I made those notes is because I had stuff that needed to be communicated but didn't know where to put the relevant doc comment. For example, the discussion of `OrderedSet` usage because it's used in a few places. I'll look over it again and see if I can find better places to put some of the information. LMK if you have any suggestions.

 > What does it mean to go offline?

Leave the pool of candidates so cannot be selected as a validator, but still bonded and can rejoin the candidate pool by calling `go_online`. It also changes their status to offline so any changes to the nominations or the validator's bond do not require updating the `CandidatePool`. This is why there are a few places with code like

```rust
if state.is_active() {
 Self::update_active_candidate()
}
```

> Design question: Can one account be a validator and also a nominator? If so can the account self nominate?

The validator and nominator roles are separate and no account is allowed to occupy both roles. So, when an account tries to `join_candidates`, there are two checks that (1) it isn't a nominator and (2) it isn't a validator candidate. For `join_nominators`, it has the same two checks.

So no account can self nominate. The validator has to be bonded > T::MinValidatorStk, but that value is stored separately from its nominations.

> Why do we need seperate bond_more (same question about bond_less) methods for validators and nominators? Maybe you could have a single fn bond_more(origin, candidate, more: BalanceOf<T>) and if the candidate is the same as the caller then treat it as a candidate bonding more. Or use an option like you did for the genesis config.

I could make this change, but I like having separate runtime methods for validator candidates and nominators because the storage items are separate. 

Also I generally try to avoid using Options as input params on runtime methods when it splits the path inside the runtime method and there is not a lot of shared logic between the branches. This changes

```rust
fn validator_bond_more() {
   _validator_bond_more()
}

fn nominator_bond_more() {
   _nominator_bond_more()
}
```

to this

```rust
fn bond_more(Option) {
 if let Some(_) = Option {
    _nominator_bond_more
  } else {
    _validator_bond_more
  }
}
```

Because they have separate storage items, the branches don't share any logic so this change makes it more nested in order to combine them into a single method. IMO it makes the runtime interface more confusing, but give me some time to think about it because this is a visceral first reaction :)

> Concern: This pallet just ticked past 1k lines code, and some parts are deeply nested. I'm reassured by the thorough test coverage. But I wonder how this footprint compares to eg pallet staking.

Still ~2500 loc less than pallet-staking/src/lib.rs and that doesn't count the other files in that pallet.

When I implemented support for multiple nominations, I added another Vec that has to be iterated over whenever we add nominations, remove nominations, or change nominations. I think it's worth it, but agree that surface area has increased.

JoshOrndorff commented:
Thanks for the answers. That all makes good sense to me.

I think it makes sense to do another learning session where the team gets to stake and validate and nominate so we can develop a stronger intuition for the various code paths. This pallet is really nice.

JoshOrndorff reviewed (approved):
LGTM
